appinsights/client.go

```diff
@@ -2,123 +2,129 @@ package appinsights
+		IsEnabled:              true,
 	}
 }
 
-func (tc *telemetryClient) Context() TelemetryContext {
-	return tc.context
-}
-
-func (tc *telemetryClient) InstrumentationKey() string {
+func (tc TelemetryClient) InstrumentationKey() string {
 	return tc.TelemetryConfiguration.InstrumentationKey
 }
 
-func (tc *telemetryClient) IsEnabled() bool {
-	return tc.isEnabled
-}
-
-func (tc *telemetryClient) SetIsEnabled(isEnabled bool) {
-	tc.isEnabled = isEnabled
-}
-
-func (tc *telemetryClient) Track(item Telemetry) {
-	if tc.isEnabled {
-		iKey := tc.context.InstrumentationKey()
-		if len(iKey) == 0 {
-			iKey = tc.TelemetryConfiguration.InstrumentationKey
-		}
+func (tc *TelemetryClient) Track(item Telemetry) {
+	if !tc.IsEnabled {
+		return
+	}
 
-		itemContext := item.Context().(*telemetryContext)
-		itemContext.iKey = iKey
+	iKey := tc.Context.InstrumentationKey
+	if len(iKey) == 0 {
+		iKey = tc.TelemetryConfiguration.InstrumentationKey
+	}
 
-		clientContext := tc.context.(*telemetryContext)
+	item.Context.InstrumentationKey = iKey
 
-		for tagkey, tagval := range clientContext.tags {
-			if itemContext.tags[tagkey] == "" {
-				itemContext.tags[tagkey] = tagval
-			}
+	for tagkey, tagval := range tc.Context.Tags {
+		if item.Context.Tags[tagkey] == "" {
+			item.Context.Tags[tagkey] = tagval
 		}
-
-		tc.channel.Send(item)
 	}
-}
-
-func (tc *telemetryClient) TrackEvent(name string) {
-	item := NewEventTelemetry(name)
-	tc.TrackEventTelemetry(item)
-}
-
-func (tc *telemetryClient) TrackEventTelemetry(event *EventTelemetry) {
-	var item Telemetry
-	item = event
-
-	tc.Track(item)
-}
-
-func (tc *telemetryClient) TrackMetric(name string, value float32) {
-	item := NewMetricTelemetry(name, value)
-	tc.TrackMetricTelemetry(item)
-}
-
-func (tc *telemetryClient) TrackMetricTelemetry(metric *MetricTelemetry) {
-	var item Telemetry
-	item = metric
-
-	tc.Track(item)
-}
-
-func (tc *telemetryClient) TrackTrace(message string) {
-	item := NewTraceTelemetry(message, Information)
-	tc.TrackTraceTelemetry(item)
-}
-
-func (tc *telemetryClient) TrackTraceTelemetry(trace *TraceTelemetry) {
-	var item Telemetry
-	item = trace
-
-	tc.Track(item)
-}
-
-func (tc *telemetryClient) TrackRequest(name string, timestamp time.Time, duration time.Duration, responseCode string, success bool) {
-	item := NewRequestTelemetry(name, timestamp, duration, responseCode, success)
-	tc.TrackRequestTelemetry(item)
-}
-
-func (tc *telemetryClient) TrackRequestTelemetry(request *RequestTelemetry) {
-	var item Telemetry
-	item = request
 
-	tc.Track(item)
+	tc.Channel.Send(item)
+}
+
+func (tc *TelemetryClient) TrackEvent(name string) {
+	tc.Track(NewEventTelemetry(name))
+}
+
+func (tc *TelemetryClient) TrackMetric(name string, value float32) {
+	tc.Track(NewMetricTelemetry(name, value))
+}
+
+func (tc *TelemetryClient) TrackTrace(
+	level SeverityLevel,
+	message string,
+	properties map[string]string,
+	alterTelementryContext func(*TelemetryContext)) {
```

Comment:
1. Spelling, should be: alterTelemetryContext
2. I don't think this belongs here, as a parameter to the Track functions.  I *do* like the idea, but I think it belongs on the TelemetryClient, perhaps as a field in the TelemetryConfiguration -- you're likely tagging *each* telemetry item.  If you're not, you can specify the right tags by modifying the telemetry after the fact.